lib/elastic_apm/basic_object.rb

```diff
@@ -0,0 +1,74 @@
+# Licensed to Elasticsearch B.V. under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# frozen_string_literal: true
+
+module ElasticAPM
```

Comment:
I though it would be nice to have a common way to create all these little objects that mostly just correspond to API objects, like `{ cloud: region }` or `{ service: name }`.

This is already going a little bit into comlex territory adding `#empty?` and `optional: true` capabilities but that is sort of a general concept for all these fields. Since we don't strip all `nil` values when generating the JSON, I thought that a nice place to put the knowledge of whether an object was _empty_ was inside its own defining class.

`BasicObject` is a loaded term though and probably not the best to use, but I struggled with coming up with something better and didn't want to think too long about it. Alternate names very welcome.

Reply:
Alternate suggestions: `APIObject`, `IntakeObject`, `EventObject`